lib/recursive-methods.rb

```diff
@@ -1,49 +1,125 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
-
-# Time complexity: ?
-# Space complexity: ?
-def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+require "pry"
+# Time complexity: O(n)
+# Space complexity: O(n)
+def factorial(num)
+  raise ArgumentError if num < 0
+  if num == 1 || num == 0 # base case
+    return 1
+  else
+    return num * factorial(num - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# O(n) for both T & S it's decrementing by 1, the function calls are the same for both
+
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length == 0 || s.length == 1 # base case
+    return s
+  end
+  return reverse(s[1..-1]) + s[0]
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  return reverse_helper(s, 0, s.length - 1)
+end
+
+def reverse_helper(s, l, r)
+  if l >= r
+    return s
+  else
+    s[l], s[r] = s[r], s[l]
+    return reverse_helper(s, l + 1, r - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError if n < 0
+  if n == 0 # base case
+    return 0
+  else
+    return bunny(n - 1) + 2
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: 0(n) look it up
 def nested(s)
```

Comment:
Similar to the above reverse this creates a new array so it's O(n^2) in time and space.